downloads/swan-lake-release-notes/swan-lake-2201.4.0/RELEASE_NOTE.md

```diff
@@ -57,6 +57,62 @@ Annotations are now supported on tuple members. Tuple members can be annotated u
 
 To view other bug fixes, see the [GitHub milestone for Swan Lake 2201.4.0](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+label%3ATeam%2FCompilerFE+milestone%3A2201.4.0+is%3Aclosed+label%3AType%2FBug).
 
+#### Backward incompatible changes
+
+- Fixed a bug that incorrectly resolved the result type of a query action that would complete normally to `error?` instead of `()`.
+  Now, resolution of the result type of a query action has been improved so that it is now set to `error` only in instances where an error can be generated during the execution of the query pipeline (`from-clause`/`join-clause`).
+
+```ballerina
+public function main() returns error? {
+    from int i in 1 ... 3
+    do {
+    };
+
+    check from int i in 1 ... 3 // warning: invalid usage of the 'check' expression operator: no expression type is equivalent to error type
+        do {
+        };
+}
+
+function iterateStream(stream<int, error?> numberStream) returns error? {
+    check from int num in numberStream
+        do {
+        };
+}
+```
+
+- Fixed a bug of propagating errors thrown from the `do-clause` in a query action to the result of the query action.
+  Now, if the execution of a statement within a `do-clause` fails with an error, it will be propagated to the nearest enclosing failure-handling statement.
```

Comment:
```suggestion
- Fixed a bug that resulted in errors returned from a `do` clause of a query action not being propagated correctly to  the result of the query action. Now, if the execution of a statement within a `do` clause fails with an error, it will be propagated to the nearest enclosing failure-handling statement.
```

Reply:
The idea changed with your above suggestion. Hope this is fine.
```suggestion
- Fixed a bug that incorrectly propagated errors returned from a `do` clause of a query action to the result of the query action. Now, if the execution of a statement within a `do` clause fails with an error, it will be propagated to the nearest enclosing failure-handling statement.
```

Reply:
Ah yes, misunderstood. LGTM.